### PR TITLE
ipa-backup: Make sure all roles are installed on the current master.

### PR DESCRIPTION
**Pull request by fcami**
ipa-backup does not check whether the IPA master it is running on has
all used roles installed. This can lead into situations where backups
are done on a CAless or KRAless host when these roles are used in the
IPA cluster. These backups cannot be used to restore a complete cluster.

With this change, ipa-backup checks what roles are installed on the
current host and compares the resulting list to the list of used roles
in the cluster. A --disable-role-check knob is also provided to restore
the previous behavior.

Fixes: https://pagure.io/freeipa/issue/8217
Signed-off-by: François Cami <fcami@redhat.com>

### CODE REVIEW

**Comment by fcami** [comment]
Marked WIP while I work on adding a proper integration test.

File: ipatests/test_integration/test_backup_and_restore.py [hunk 1]
```diff
@@ -909,3 +916,122 @@ def test_restore_trust_pkg_before_restore(self):
             tasks.install_packages(self.master, ['*ipa-server-trust-ad'])
             self.master.run_command(['ipa-restore', backup_path],
                                     stdin_text=dirman_password + '\nyes')
+
+
+class TestBackupRoles(IntegrationTest):
+    """ipa-backup should only run on replicas with all the in-use roles"""
```
Comment (miskopo) [inline ipatests/test_integration/test_backup_and_restore.py]: Docstrings enclosed in """ are multiline-safe, there is no need for splitting over next line.

**Comment by miskopo** [comment]
Please, see my in-line comment. Apart from that, the test looks good to me.

File: ipatests/test_integration/test_backup_and_restore.py (hunk 1 above)
Comment (fcami) [reply ipatests/test_integration/test_backup_and_restore.py]: Thanks, I've fixed it.

**Comment by fcami** [comment]
Main run: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/694efee2-7037-11ea-a07d-fa163e23289d/

**Comment by fcami** [comment]
Successful run: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/694efee2-7037-11ea-a07d-fa163e23289d/
Squashing unimportant commits.

**Comment by fcami** [comment]
Successful run: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/fa57beb0-7046-11ea-a1c5-fa163e2d2908/

**Comment by fcami** [comment]
hi @mrizwan93 for readability and documentation purposes yes, thanks for the suggestion. Done.

File: ipaserver/install/ipa_backup.py [hunk 2]
```diff
@@ -358,6 +363,79 @@ def run(self):
...
+        # Controllers are Trust Agents themselves and contain extra,
+        # necessary Samba configuration. So either the cluster has no
+        # AD Trust bits installed, or it should be backuped on a Trust
+        # Controller, not a Trust Agent.
+        role_names = {
+            'CA', 'DNS', 'DNSKeySync', 'KRA', 'ADTRUST'
+        }
+
+        search_base = DN(api.env.container_masters, api.env.basedn)
+        attrs_list = ['ipaConfigString', 'cn']
```
Comment (rcritten) [inline ipaserver/install/ipa_backup.py]: Please make the attributes all lower-case.

Comment (fcami) [reply ipaserver/install/ipa_backup.py]: I think this is resolved now.

**Comment by rcritten** [comment]
Looks good, please remove temp commits.

**Comment by fcami** [comment]
Hi @rcritten thanks, the temp commit is removed.

**Comment by fcami** [comment]
master:

* 3a9b66b5302903f4abe75c3ee93b457f5c1dc405 ipatests: test ipa-backup with different role configurations.
* 9324bba6b7db4152ff85017d59000fc204ce2c90 test_backup_and_restore: add server role verification steps
* 3665ba928b25c017465588bd505e93d15d72290f ipa-backup: Make sure all roles are installed on the current master.

**Pull request closed without merging**